data/tutorials/language/0lg_10_polymorphic_variants.md

```diff
@@ -0,0 +1,713 @@
+
+Here is an equivalent version using polymorphic variants:
+```ocaml
+# let ( let* ) = Result.bind;;
+val ( let* ) : ('a, 'b) result -> ('a -> ('c, 'b) result) -> ('c, 'b) result =
+  <fun>
+
+# let init n f = try Ok (List.init n f) with
+    | Invalid_argument _ -> Error `Negative_length;;
+val init : int -> (int -> 'a) -> ('a list, [> `Negative_length ]) result =
+
+# let nth u i = try Ok (List.nth u i) with
+    | Invalid_argument _ -> Error (`Negative_index i)
+    | Failure _ -> Error (`Too_short (List.length u, i));;
+val nth :
+  'a list ->
+  int -> ('a, [> `Negative_index of int | `Too_short of int * int ]) result =
+  <fun>
+
+# let f_res m n =
+    let* u = init m Fun.id in
+    let u = List.map (fun n -> n * n) u in
+    let* x = nth u n in
+    Ok x;;
+val f_res :
+  int ->
+  int ->
+  (int,
+   [> `Negative_index of int | `Negative_length | `Too_short of int * int ])
+  result = <fun>
+
+```
+
+Using polymorphic variants, the type-checker generates a unique type for the whole pipe. The constraints coming from calling `init` and `nth_error` are merged into a single type.
+
+## Drawbacks
+
+### Hard-to-Read Type Expressions
+
+By default, polymorphic variant types aren't declared with a name before being used. The compiler will generate type expressions corresponding to each function dealing with polymorphic variants. Some of those types are hard to read. When several such functions are composed together, inferred types can become very large and difficult to understand.
+```ocaml
+# let rec fold_left f y = function `Nil -> y | `Cons (x, u) -> fold_left f (f x y) u;;
+val fold_left :
+  ('a -> 'b -> 'b) -> 'b -> ([< `Cons of 'a * 'c | `Nil ] as 'c) -> 'b = <fun>
+```
+
+### Over Constrained Types
+
+In some circumstances, combining sets of constraints will artificially reduce the domain of a function. This happens when the conjunction of constraints must be taken. Functions `f` and `g` used here are those defined in the [first section](#a-first-example).
+```ocaml
+# f;;
+- : [< `Broccoli | `Fruit of string ] -> string = <fun>
+
+# g;;
+- : [< `Broccoli | `Edible of string ] -> string = <fun>
+
+# let u = [f; g];;
+u : ([< `Broccoli ] -> string) list = [<fun>; <fun>]
+
+# f (`Fruit "Pitahaya");;
+- : string = "Pitahaya"
+
+# (List.hd u) (`Fruit "Pitahaya");;
+Error: This expression has type [> `Fruit of string ]
+       but an expression was expected of type [< `Broccoli ]
+       The second variant type does not allow tag(s) `Fruit
+```
+
+Function `f` accepts tags `` `Broccoli`` and `` `Fruit`` whilst `g` accepts `` `Broccoli`` and `` `Edible``. But if `f` and `g` are stored in a list, they must have the same type. That forces their domain to be restricted to a common subtype. Although `f` can handle the tag `` `Fruit``, it no longer accepts that parameter when `f` is extracted from the list.
+
+### Weaken Type-Checking and Harder Debugging
```

Comment:
When you write "weaken type-checking" are you referring to having both a `Gray` and `Grey` tags? I assumed this was meant to be an example of an accident/typo, but I don't think its fair to call it "weakened type-checking".

Like the compiler isn't going to mistake `Grey` for `Gray`.

Reply:
> Like the compiler isn't going to mistake Grey for Gray.

But would if it had been simple variants. That's the point. But I didn't say it explicitly, so that's on me.